File: hackertracker/src/main/java/com/shortstack/hackertracker/Database/DatabaseController.java
package com.shortstack.hackertracker.Database;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteException;
import android.support.annotation.NonNull;

import com.google.gson.FieldNamingPolicy;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.orhanobut.logger.Logger;
import com.shortstack.hackertracker.Application.App;
import com.shortstack.hackertracker.Common.Constants;
import com.shortstack.hackertracker.Event.FavoriteEvent;
import com.shortstack.hackertracker.Model.Company;
import com.shortstack.hackertracker.Model.Item;
import com.shortstack.hackertracker.Utils.NotificationHelper;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Calendar;
import java.util.List;

public class DatabaseController {

    private static final String SCHEDULE = "hackertracker.sqlite";
    private static final String VENDORS = "vendors.sqlite";


    private static final int SCHEDULE_VERSION = 210;
    private static final int VENDOR_VERSION = 14;
    public static final String SCHEDULE_TABLE_NAME = "data";

    private SQLiteDatabase mSchedule;
    private SQLiteDatabase mVendors;

    private Gson mGson;

    public DatabaseController(Context context) {
        DatabaseHelper databaseAdapter;

        databaseAdapter = new DatabaseHelper(context, SCHEDULE, SCHEDULE_VERSION);
        mSchedule = databaseAdapter.getWritableDatabase();

        databaseAdapter = new DatabaseHelper(context, VENDORS, VENDOR_VERSION);
        mVendors = databaseAdapter.getWritableDatabase();

        mGson = new GsonBuilder().setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES).create();

//        Logger.d("Version: " + mSchedule.getVersion());
//        Logger.d("Version: " + mVendors.getVersion());


//        checkBookmark();
    }


    @Override
    protected void finalize() throws Throwable {
        super.finalize();
        mSchedule.close();
        mVendors.close();
    }

    public void toggleBookmark(Item item) {
        int value = item.isBookmarked() ? Constants.UNBOOKMARKED : Constants.BOOKMARKED;
        setScheduleBookmarked(value, item.getId());

        item.toggleBookmark();
        App.Companion.getApplication().postBusEvent(new FavoriteEvent(item.getId()));

        if (item.isBookmarked()) {
            App.Companion.getApplication().getNotificationHelper().scheduleItemNotification(item);
        }
    }

    private void setScheduleBookmarked(int state, int id) {
        mSchedule.execSQL("UPDATE data SET starred=" + state + " WHERE id=" + id);
    }

    public void addScheduleItem(Item item) {
        mSchedule.insert(SCHEDULE_TABLE_NAME, null, item.getContentValues(mGson));

        Logger.d("Inserted item.");

        Cursor cursor = mSchedule.rawQuery("SELECT * FROM data", new String[]{});
        int count = cursor.getCount();
        Logger.d("Count now: " + count);


        cursor.moveToFirst();

        do{


        } while(cursor.moveToNext());



        cursor.close();
    }

    public void updateScheduleItem(Item item) {
        String filter = "id=?";
        String[] args = new String[]{String.valueOf(item.getId())};

        SQLiteDatabase schedule = App.Companion.getApplication().getDatabaseController().getSchedule();
        ContentValues values = item.getContentValues(mGson);
        values.put("is_new", "1");

        int rowsUpdated = schedule.update(SCHEDULE_TABLE_NAME, values, filter, args);
        if( rowsUpdated == 0 ) {
            schedule.insert(SCHEDULE_TABLE_NAME, null, values);
            // Change to insert new event.
//            App.Companion.getApplication().postBusEvent(new UpdateListContentsEvent());
        } else {
            // Change to update event.
//            App.Companion.getApplication().postBusEvent(new UpdateListContentsEvent());





            if( item.getId() == 7 ) {
                Logger.d("Updating Cafe Bar.  " + item.isBookmarked());

                checkBookmark();


            }

            //long time = System.currentTimeMillis();

            Item item1 = getScheduleItem(item.getId());

            //Logger.d("Time to fetch: " + (System.currentTimeMillis() - time));

            if( item1.isBookmarked() ) {

                NotificationHelper notificationHelper = App.Companion.getApplication().getNotificationHelper();
                // Cancel the notification, in case the time changes.
                notificationHelper.cancelNotification( item.getId() );

                // Set a new one.
                notificationHelper.scheduleItemNotification(item);

                notificationHelper.postNotification( notificationHelper.getUpdatedItemNotification(item), item.getId());
            }
        }
    }

    public void updateSchedule(List<Item> items) {
        for (Item item : items) {
            updateScheduleItem(item);
        }
    }

    public Item getScheduleItem( int id ) {


        Cursor cursor = mSchedule.query("data", null, "id=?", new String[]{String.valueOf(id)}, null, null, null, null);

        if( cursor.moveToFirst() ) {
            Item item = Item.CursorToItem(mGson, cursor);
            cursor.close();



            return item;
        }

        return new Item();

        //throw new IllegalStateException("Could not find schedule item with id=" + id);
    }

    public void checkBookmark() {
        Cursor cursor = mSchedule.query("data", null, "id=?", new String[]{String.valueOf(14)}, null, null, null, null);

        Logger.d("Cursor length: " + cursor.getCount());
        cursor.moveToFirst();
        Item item1 = Item.CursorToItem(mGson, cursor);

        Logger.d("Item: " + item1.getTitle() + " " + item1.isBookmarked());

        cursor.close();
    }

    public SQLiteDatabase getSchedule() {
        return mSchedule;
    }

    public List<Item> getItemByDate(String... type) throws SQLiteException {
        ArrayList<Item> result = new ArrayList<>();
        ArrayList<String> args = new ArrayList<>(Arrays.asList(type));


        long time;


        // Types
        String selection = "(type=?";
        for (int i = 0; i < type.length - 1; i++) {
            selection = selection.concat(" OR type=?");

        }
        selection = selection.concat(")");

        // Date
        if( App.Companion.getStorage().showActiveEventsOnly()) {
            Calendar currentDate = App.Companion.getCurrentCalendar();
            SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd");
            SimpleDateFormat timeFormat = new SimpleDateFormat("hh:mm");

            selection = selection.concat("AND (( date = ? AND end > ? )");

            args.add(dateFormat.format(currentDate.getTime()));
            args.add(timeFormat.format(currentDate.getTime()));

            // Next Day
            if( true ) {
                currentDate.roll(Calendar.DAY_OF_YEAR, true);
                // All next day
                selection = selection.concat(" OR date = ?");
                args.add(dateFormat.format(currentDate.getTime()));

                // Any days after it.
                selection = selection.concat(" OR date > ?");
                args.add(dateFormat.format(currentDate.getTime()));

            }

            selection = selection.concat(")");
        }


        // Debugging
        String args_print = "";
        for (String arg : args) {
            args_print = args_print.concat(arg + ", ");
        }

        Logger.d("Selection: " + selection + " Args: " + args_print);
        time = System.currentTimeMillis();


        // Query
        Cursor cursor = mSchedule.query(SCHEDULE_TABLE_NAME, null, selection, args.toArray(new String[args.size()]), null, null, "date, begin");

        Logger.d("Cursor: " + cursor.getCount() + " Took: " + (System.currentTimeMillis() - time));


        time = System.currentTimeMillis();

        // Adding to list
        if( cursor.moveToFirst() ) {

            do {
                 result.add(Item.CursorToItem(mGson, cursor));
            } while (cursor.moveToNext());

        }
        cursor.close();

        Logger.d("Size of Results: " + result.size() + " Took: " + (System.currentTimeMillis() - time));

        return result;
    }


    @NonNull
    private String getQueryString(String[] type) {
        String query = "SELECT * FROM data WHERE";
        if (type.length > 0) {
            query = query.concat(" (");
            for (int i = 0; i < type.length; i++) {
                query = query.concat("type=?");
                if (i < type.length - 1) query = query.concat(" OR ");
            }
            query = query.concat(") OR");
        }
        query = query.concat(" starred=1 ORDER BY date, begin");
        return query;
    }

    public List<Company> getVendors() {
        ArrayList<Company> result = new ArrayList<>();

        Cursor cursor = mVendors.rawQuery("SELECT * FROM data", new String[]{});

        try {
            if (cursor.moveToFirst()) {
                do {

                    Company item = new Company();

                    item.setId(cursor.getInt(cursor.getColumnIndex("id")));
                    item.setTitle(cursor.getString(cursor.getColumnIndex("title")));
                    item.setDescription(cursor.getString(cursor.getColumnIndex("description")));
                    item.setLink(cursor.getString(cursor.getColumnIndex("link")));
                    item.setPartner(cursor.getInt(cursor.getColumnIndex("partner")));

                    result.add(item);
                } while (cursor.moveToNext());
            }
        } finally {
            cursor.close();
        }

        return result;
    }
}

File: hackertracker/src/main/java/com/shortstack/hackertracker/Database/DatabaseController.kt
package com.shortstack.hackertracker.database

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import com.google.gson.FieldNamingPolicy
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.orhanobut.logger.Logger
import com.shortstack.hackertracker.App
import com.shortstack.hackertracker.BuildConfig
import com.shortstack.hackertracker.models.Item
import com.shortstack.hackertracker.joinSQLOr
import com.shortstack.hackertracker.utils.SharedPreferencesUtil
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.io.IOException

abstract class DatabaseController(protected val context: Context, protected val storage: SharedPreferencesUtil, name: String, version: Int) : SQLiteOpenHelper(context, name, null, version) {


    val gson = GsonBuilder().setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES).create()

    // SQL
    val SELECT_ALL_FROM = "SELECT * from "
    val LIKE = "LIKE ?"
    val DATABASE_DIRECTORY = "database"

    fun checkDatabase() {
        val db = writableDatabase

        if (isScheduleOutOfDate()) {
            Logger.d("Have not synced the current version. Update the database with the json.")
            updateDataSet(db)
        }
    }

    fun exists() = DatabaseController.exists(context, databaseName)

    private fun isScheduleOutOfDate() = storage.lastSyncVersion != BuildConfig.VERSION_CODE/* || BuildConfig.DEBUG*/

    protected fun finalize() {
        close()
    }

    override fun onCreate(db: SQLiteDatabase) {
        initDatabase(db, gson)
    }

    abstract protected fun initDatabase(db: SQLiteDatabase, gson: Gson)
    abstract protected fun initOtherDatabases(gson: Gson, db: SQLiteDatabase)
    abstract protected fun updateDatabase(gson: Gson, db: SQLiteDatabase)

    fun updateDataSet(db: SQLiteDatabase) {
        updateDatabase(gson, db)
    }


    protected fun applyPatches(database: SQLiteDatabase, patches: Patches) {
        patches.patches.indices
                .map { patches.patches[it] }
                .forEach {
                    for (command in it.commands) {
                        database.execSQL(command)
                    }
                }
    }

    protected fun <T> initTable(database: SQLiteDatabase, table: String, array: Array<T>) {
        database.beginTransaction()

        // Clearing out the table.
        database.delete(table, null, null)
        try {
            array.forEach {
                val values = getContentValues(it, gson)
                database.insert(table, null, values)
            }
            database.setTransactionSuccessful()
        } finally {
            database.endTransaction()
        }
    }

    fun <T> query(table: String, t: Class<T>,
                  columns: Array<String>? = null, selection: String? = null, selectionArgs: MutableList<String>? = null,
                  groupBy: String? = null, having: String? = null, orderBy: String? = null, limit: Int? = null, page: Int = 0,
                  searchColumns: MutableList<String>? = null, searchText: String? = null
    ): ArrayList<T> {
        val result = ArrayList<T>()
        var limitOffset: String? = null
        var finalSelection = selection
        var finalSelectArgs = selectionArgs

        // Limit and offset
        if (limit != null) {
            limitOffset = "${page * limit}, $limit"
        }


        // Search by text
        if (searchText != null) {

            val columns = getSearchColumns(searchColumns!!)?.joinSQLOr()

            if( finalSelection == null ) {
                finalSelection = columns
            } else {
                finalSelection += columns
            }

            val array = Array<String>(searchColumns.size, { "%$searchText%" })

            if (finalSelectArgs == null)
                finalSelectArgs = array.toMutableList()
            else
                finalSelectArgs.addAll(array)
        }


        val cursor = readableDatabase.query(table, columns, finalSelection, finalSelectArgs?.toTypedArray(), groupBy, having, orderBy, limitOffset)

        if (cursor.moveToFirst()) {
            do {
                result.add(fromCursor(t, cursor))
            } while (cursor.moveToNext())
        }
        cursor.close()

        return result
    }

    fun getSearchColumns( searchColumns: MutableList<String> ) : Array<String>? {
        return Array( searchColumns.size, {
            "${searchColumns[it]} $LIKE"
        } )
    }

    protected fun getJSONFromFile(filename: String): String {
        try {
            val s = "$DATABASE_DIRECTORY/$databaseName/$filename"
            val stream = context.assets.open(s)

            val size = stream.available()

            val buffer = ByteArray(size)

            stream.read(buffer)
            stream.close()

            return String(buffer)

        } catch (e: IOException) {
            Logger.e(e, "Could not create the database.")
        }

        return ""
    }


    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        TODO()
    }

    fun <T> fromCursor(t: Class<T>, cursor: Cursor): T {
        val obj = getJsonObject(cursor)

        if (t == Item::class.java) {
            val who = obj.getString("who")
            obj.remove("who")
            obj.put("who", JSONArray(who))
        }

        return gson.fromJson(obj.toString(), t)
    }


    companion object {
        fun exists(context: Context, database: String): Boolean {
            val dbFile = context.getDatabasePath(database)
            return dbFile.exists()
        }



        private fun getJsonObject(cursor: Cursor): JSONObject {
            val obj = JSONObject()

            val totalColumn = cursor.columnCount

            for (i in 0..totalColumn - 1) {
                try {
                    obj.put(cursor.getColumnName(i), cursor.getString(i))
                } catch (e: Exception) {
                    Logger.e(e, "Failed to convert Cursor into JSONObject.")
                }
            }
            return obj
        }

        fun <T> getContentValues(t: T, gson: Gson): ContentValues {
            val values = ContentValues()

            val json = gson.toJson(t)
            try {
                val obj = JSONObject(json)

                val keys = obj.keys()
                var key: String
                while (keys.hasNext()) {
                    key = keys.next()
                    values.put(key, obj.getString(key))
                }

            } catch (e: JSONException) {
                e.printStackTrace()
            }

            return values
        }
    }


}
